lib/pbench/server/api/resources/query_apis/__init__.py

```diff
@@ -1,94 +1,510 @@
+            name: Parameter name
+            type: Parameter type
+            required: whether the parameter is required (default to False)
+        """
+        self.name = name
+        self.type = type
+        self.required = required
+
+    def invalid(self, json: Dict[AnyStr, Any]) -> bool:
+        """
+        ok Check whether the expression of this parameter in the JSON document
+        is invalid.
+
+        If the parameter is required and missing (or null), then it's invalid;
+        if the parameter is not required and specified as null, it's invalid.
+
+        Args:
+            json: The client JSON document being validated.
+
+        Returns:
+            True if unacceptable
+        """
+        return (self.required and self.name not in json) or (
+            self.name in json and not json[self.name]
+        )
+
+    def __str__(self) -> str:
+        return f"Parameter<{self.name}:{self.type}>"
+
+
+class Schema:
+    """
+    Define the client input schema for a server query that's based on
+    Elasticsearch.
+    """
+
+    def __init__(self, *parameters: Parameter):
+        """
+        __init__ Specify an interface schema as a list of Parameter objects.
+
+        Args:
+            parameters: a list of Parameter objects
+        """
+        self.parameters = {p.name: p for p in parameters}
+
+    def get(self, name: AnyStr) -> Parameter:
+        """
+        get Retrieve a parameter by name
+
+        Args:
+            name: Parameter name to return
+
+        Raises:
+            KeyError: The parameter wasn't found
+
+        Returns:
+            the specified Parameter value
+        """
+        if name not in self.parameters:
+            raise KeyError(f"Parameter {name} not found")
+        return self.parameters[name]
+
+    def __iter__(self) -> Iterable[Parameter]:
+        """
+        __iter__ Iterating on a Schema iterates through the list of parameters
+
+        Returns:
+            Parameter list iterator
+        """
+        return iter(self.parameters.items())
+
+    def validate(self, json_data: Dict[AnyStr, Any]) -> Dict[AnyStr, Any]:
+        """
+        validate Validate an incoming JSON document against the schema and
+        return a new JSON dict with translated values.
+
+        Args:
+            json_data: Incoming client JSON document
+
+        Returns:
+            New JSON document with validated and possibly translated values
+        """
+        if not json_data:
+            raise InvalidRequestPayload()
+
+        bad_keys = [n for n, p in self.parameters.items() if p.invalid(json_data)]
+        if len(bad_keys) > 0:
+            raise MissingParameters(bad_keys)
+
+        processed = {}
+        for p in json_data:
+            tp = self.parameters.get(p)
+            original = json_data[p]
+            if tp:
+                value = tp.type.convert(original)
+                processed[p] = value
+            else:
+                processed[p] = original
```

Comment:
Consider using a conditional, instead:
```
            processed[p] = tp.type.convert(json_data[p]) if tp else json_data[p]
```
That gets rid of two local variables of very limited value.

Reply:
It felt more clear expanded when I wrote it; but this actually isn't any worse to read, so I clearly blew that.